plugins/com.google.cloud.tools.eclipse.projectselector.test/src/com/google/cloud/tools/eclipse/projectselector/MiniSelectorTest.java

```diff
@@ -89,31 +101,45 @@ public void testListenerFired() {
 
     assertNotNull(selector.getSelection());
     assertTrue(selector.getSelection().isEmpty());
-    final boolean[] calledAndCorrect = new boolean[] {false};
+    final AtomicReference<ISelection> lastSelection = new AtomicReference<>();
     selector.addSelectionChangedListener(new ISelectionChangedListener() {
       @Override
       public void selectionChanged(SelectionChangedEvent event) {
-        assertThat(event.getSelection(), instanceOf(IStructuredSelection.class));
-        IStructuredSelection selection = (IStructuredSelection) event.getSelection();
-        assertEquals(1, selection.size());
-        assertThat(selection.getFirstElement(), instanceOf(GcpProject.class));
-        GcpProject gcpProject = (GcpProject) selection.getFirstElement();
-        calledAndCorrect[0] =
-            "foo".equals(gcpProject.getName()) && "foo.id".equals(gcpProject.getId());
+        lastSelection.set(event.getSelection());
       }
     });
-    assertFalse(calledAndCorrect[0]);
 
     selector.setProject("foo.id");
-    spinEvents(); // setProject() waits for the project list to be returned
+    waitUntilResolvedProject(selector); // waits for the project list to be returned
 
-    assertTrue(calledAndCorrect[0]);
+    assertThat(lastSelection.get(), instanceOf(IStructuredSelection.class));
+    IStructuredSelection selection = (IStructuredSelection) lastSelection.get();
+    assertEquals(1, selection.size());
+    assertThat(selection.getFirstElement(), instanceOf(GcpProject.class));
+    GcpProject gcpProject = (GcpProject) selection.getFirstElement();
+    assertEquals("foo", gcpProject.getName());
+    assertEquals("foo.id", gcpProject.getId());
   }
 
-  private void spinEvents() {
-    while (Display.getCurrent().readAndDispatch());
+  private void waitUntilResolvedProject(final MiniSelector selector) {
+    bot.waitUntil(new DefaultCondition() {
+      @Override
+      public boolean test() throws Exception {
+        if (Display.getCurrent() != null) {
+          // seems surprising that this is required?
```

Comment:
I looked at the `waitUtil()` code, and it just repeatedly calls `Thread.sleep()`. It were fine if the tests were running in a non-UI thread like our existing integration test bundle does, but since we are in the UI thread, this is certainly necessary.

Reply:
But it also means that the SWT event loop isn't processed when this is the UI thread.

Reply:
> But it also means that the SWT event loop isn't processed when this is the UI thread.

No idea if the SWT event loop will be executing in another thread, but what is certain is that `waitUtil()` is just a loop with `Thread.sleep()` and `ICondition.test()`.

```java
	public void waitUntil(ICondition condition, ...) ... {
		while (true) {
			...
				if (condition.test()) return;
			...
			sleep(interval);  // This is just `Thread.sleep()`.
		}
	}
```

Reply:
Ah sorry, my comment was meant as a warning for future readers as to why this was required.  I'll make it clear.